File: 10-Adapter/src/cpp/main.cpp
#include <iostream>
#include <memory>
#include "StripeAdapter.h"
#include "PayPalAdapter.h"

void process_order(PaymentProcessor& processor, double amount, const std::string& currency) {
    bool success = processor.process_payment(amount, currency);
    std::cout << "  Payment " << (success ? "succeeded" : "failed") << "\n\n";
}

void refund_order(PaymentProcessor& processor, const std::string& transaction_id) {
    bool success = processor.refund(transaction_id);
    std::cout << "  Refund " << (success ? "succeeded" : "failed") << "\n\n";
}

int main() {
    std::cout << "=== Adapter Pattern Demo (C++) ===\n\n";

    StripeApi stripe_api;
    PayPalApi paypal_api;

    StripeAdapter stripe(stripe_api, "sk_test_abc123");
    PayPalAdapter paypal(paypal_api);

    // 1. Process payments through the same interface
    std::cout << "--- Process Payments ---\n";
    process_order(stripe, 49.99, "USD");
    process_order(paypal, 29.99, "EUR");

    // 2. Refunds
    std::cout << "--- Refund Payments ---\n";
    refund_order(stripe, "ch_12345");
    refund_order(paypal, "PAY-67890");

    // 3. Swap implementations
    std::cout << "--- Swap Implementation ---\n";
    PaymentProcessor* processor = &stripe;
    process_order(*processor, 99.99, "USD");

    processor = &paypal;
    process_order(*processor, 99.99, "USD");

    return 0;
}

File: 10-Adapter/src/rust/main.rs
mod payment_processor;
mod stripe_api;
mod paypal_api;
mod stripe_adapter;
mod paypal_adapter;

use payment_processor::PaymentProcessor;
use stripe_api::StripeApi;
use paypal_api::PayPalApi;
use stripe_adapter::StripeAdapter;
use paypal_adapter::PayPalAdapter;

fn process_order(processor: &dyn PaymentProcessor, amount: f64, currency: &str) {
    let success = processor.process_payment(amount, currency);
    println!("  Payment {}\n", if success { "succeeded" } else { "failed" });
}

fn refund_order(processor: &dyn PaymentProcessor, transaction_id: &str) {
    let success = processor.refund(transaction_id);
    println!("  Refund {}\n", if success { "succeeded" } else { "failed" });
}

fn main() {
    println!("=== Adapter Pattern Demo (Rust) ===\n");

    let stripe = StripeAdapter::new(StripeApi, "sk_test_abc123");
    let paypal = PayPalAdapter::new(PayPalApi);

    // 1. Process payments through the same trait
    println!("--- Process Payments ---");
    process_order(&stripe, 49.99, "USD");
    process_order(&paypal, 29.99, "EUR");

    // 2. Refunds
    println!("--- Refund Payments ---");
    refund_order(&stripe, "ch_12345");
    refund_order(&paypal, "PAY-67890");

    // 3. Swap implementations via trait object
    println!("--- Swap Implementation ---");
    let processor: &dyn PaymentProcessor = &stripe;
    process_order(processor, 99.99, "USD");

    let processor: &dyn PaymentProcessor = &paypal;
    process_order(processor, 99.99, "USD");
}
